indico/modules/events/contributions/templates/display/contribution_display.html

```diff
@@ -273,7 +273,13 @@ <h3>{% trans %}Subcontributions{% endtrans %}</h3>
             </div>
         </section>
     {%- endif %}
-    {% if contribution.event.has_feature('papers') %}
+
+    {% set paper_hidden = contribution.event.cfp.hide_peer_reviewed_papers %}
+    {% set permitted_hidden_paper_viewers = contribution.paper.can_judge(session.user) or contribution.paper.can_review(session.user) or contribution.paper.can_manage(session.user) or contribution.can_submit_proceedings(session.user) %}
+
+    {% if contribution.event.has_feature('papers') and not paper_hidden %}
+        {{ render_paper_section(contribution) }}
+    {% elif paper_hidden and permitted_hidden_paper_viewers %}
```

Comment:
Check the `render_paper_section` macro. It already contains some logic, and it should be much easier to extend this to take the new flag into account.